### PR TITLE
Making ops c10 full: out overloads with default arguments

### PR DESCRIPTION
**Pull request by smessmer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49259 Remove `use_c10_dispatcher: full` lines
* #49251 Remove generated_unboxing_wrappers and setManuallyBoxedKernel
* #49220 Remove .impl_UNBOXED() and functionalities associated with it
* #49164 Remove codegen logic to support non-c10-full ops
* #49138 [wip] Making ops c10-full: list of optional tensors
* #49150 Making ops c10-full: ConstQuantizerPtr
* #49146 Making ops c10-full: Storage arguments
* #49088 Making ops c10-full: optional lists
* #49083 Making ops c10 full: optional out arguments
* #49013 Making ops c10-full: Generator arguments
* **#49012 Making ops c10 full: out overloads with default arguments**
* #49008 Making ops c10-full: Dimname arguments
* #49007 Making ops c10-full: ops blocked by manual registrations
* #49006 Making ops c10 full: ops returning multiple out arguments
* #48912 Make out ops c10-full (with hacky-wrapper)

For some reason we apply default arguments to the functions in at::native too. So when an out overload had default arguments,
we couldn't move the out argument to the end because of those default arguments preceding it.
This PR fixes that and makes out overloads with default arguments c10-full

Differential Revision: [D25394605](https://our.internmc.facebook.com/intern/diff/D25394605/)

### CODE REVIEW

File: tools/codegen/gen.py
Comment (ezyang) [inline tools/codegen/gen.py:600]: In particular, direct `native::` calls rely on the defaulting quite a bit.

**Review by ezyang — APPROVED** [review]

File: tools/codegen/gen.py
Comment (bhosmer) [inline tools/codegen/gen.py:600]: Sorry if I'm being dense but won't only kernels for `c10-full` out ops have the problem described here?  IIUC these are declarations of user-supplied kernels, meaning a port to c10-full would need to take place before this situation arises - until then, the out-first signatures of the existing kernels would necessarily continue to appear here.

If so, we should restrict the default stripping to c10-full out ops, to avoid breaking existing direct calls for no reason.

Orthogonally, minor point but why not just a simple if statement that puts the two different versions of the arg list together? Avoids the nested function and reevaluating the test for every argument.

**Review by bhosmer — APPROVED** [review]
Accepting to unblock, but if I'm right re the inline comment we should really change that before landing

File: tools/codegen/gen.py
Comment (smessmer) [reply tools/codegen/gen.py:600]: Restricting it to c10 full ops would complicate the landscape because a caller now has to know if it's c10-full or not. Treating all ops the same is more consistent since c10-fullness is an implementation detail that isn't supposed to have API surface. On another note, they're going to be all c10-full soon anyways. With this PR, the overwhelming majority of out ops is already c10-full, so it probably doesn't matter. Note also that at::native isn't an API that we guarantee bc for anyways.

These are declarations of user supplied kernels but default arguments should only occur in the declaration, not in the user supplied definition of the kernel. So this shouldn't have any effect on writers of at::native kernels, only on people who call that API because they don't want to go through the dispatcher.

Comment (bhosmer) [reply tools/codegen/gen.py]: But `NativeFunctions.h` declares user-written kernels, right? 

If so, then kernels for hacky wrapper out ops don't have the issue, since their out args are still at the beginning. For those I think this change will needlessly strip defaults. Knowing whether a kernel is c10-full or not doesn't really apply - it's a matter of whether we leave existing calls undisturbed or not. 

(It's true that moving an op from hacky_wrapper to c10-full will require porting the native kernels and their callers, but I don't see the benefit in churning the callers prior to that.)

Concrete example: here's a kernel from NativeFunctions.h before this change:
```
CAFFE2_API Tensor & cumprod_out(Tensor & out, const Tensor & self, int64_t dim, c10::optional<ScalarType> dtype=c10::nullopt);
```
after:
```
CAFFE2_API Tensor & cumprod_out(Tensor & out, const Tensor & self, int64_t dim, c10::optional<ScalarType> dtype);
```

Note that this is an out op that was hacky wrapper *already*, before this PR, so clearly the default wasn't a problem.

Comment (smessmer) [reply tools/codegen/gen.py]: ok you convinced me, I'll only remove them for c10-full ops.

**Pull request closed without merging**
